### ISSUE TITLE
HMM Multivariate from samples

### ISSUE THREAD

**Issue by precupstefan** [jmschrei/pomegranate#714]
Hello.

I'm trying to create a HMM model from samples using multiple observations.

From my understanding we should pass something like this [observationA, observations, .... observationE].

and when using from samples we do 
`
 HiddenMarkovModel.from_samples(MultivariateGaussianDistribution, 8, [[[obsA,...obsE]]])
`

The problem is that sometimes the method above works, and other times I get the following error:
```
  File "pomegranate\hmm.pyx", line 3658, in pomegranate.hmm.HiddenMarkovModel.from_samples
  File "pomegranate\kmeans.pyx", line 442, in pomegranate.kmeans.Kmeans.fit
  File "pomegranate\kmeans.pyx", line 452, in pomegranate.kmeans.Kmeans.fit
  File "pomegranate\kmeans.pyx", line 133, in pomegranate.kmeans.initialize_centroids
  File "mtrand.pyx", line 922, in numpy.random.mtrand.RandomState.choice
ValueError: probabilities contain NaN
```

what an observation is composed of:
```state: int  values: [0,127]
height: int values [150 190]
sleep_quality: int values [1,5]
eyhlass_wearer int values 0 or 1
gender: int values 0 or 1
```

When using that format, HMM works every time. Because I have too few data I have to restrict height and sleep quality to 1 or 0.

```
state: int  values: [0,127]
height: int values 1 or 0
sleep_quality: int values 0 or 1
eyhlass_wearer int values 0 or 1
gender: int values 0 or 1
```

The issue appears when I'm using the simplified format. Sometimes the error above appears, other times it just works and there are times where even though the model trains, the log_probability always returns -infinity.

Can you please tell me why this is happening? Thank you

**Comment by jmschrei** [comment]
It's difficult for me to say exactly what the issue is without code to reproduce the bug but I'm guessing there's an issue with there not being enough observations to support a covariance matrix of the size you're specifying. 

The way you're supposed to format the data is a list of sequences. For instance, if you have two  3-dimensional sequence like such:

```
X1 = numpy.array([[1, 5, 3], [2, 3, 1] ,[8, 5, 2], [2, 3, 5], [2, 5, 3], [1, 2, 3], [1, 2, 3]])
X2 = numpy.array([[2, 1, 3], [2, 2, 2], [2, 1, 1], [1, 3, 4], [2, 4, 1]])
```

Then you'd train with

```
HiddenMarkovModel.from_samples(MultivariateGaussianDistribution, 8, [X1, X2])
```

Learning full covariance matrices with integer data is sometimes difficult to do. You might consider using simple normal distributions that assume independent between the features to start off with. You'd do that with 

```
HiddenMarkovModel.from_samples(NormalDistribution, 8, [X1, X2])
```

**Comment by precupstefan** [comment]
Using 
`HiddenMarkovModel.from_samples(NormalDistribution, 8, [X1, X2])` seems to train the model, but the distribution is always NAN and no log_probability can be calculated.

Here's how my training data looks like:
```
training_data = [
 [[0, 1, 1, 1, 1], [8, 1, 1, 1, 1], [24, 1, 1, 1, 1], [56, 1, 1, 1, 1], [57, 1, 1, 1, 1], [59, 1, 1, 1, 1], [63, 1, 1, 1, 1], [127, 1, 1, 1, 1]],
 [[0, 1, 1, 1, 1], [16, 1, 1, 1, 1], [48, 1, 1, 1, 1], [50, 1, 1, 1, 1], [58, 1, 1, 1, 1], [62, 1, 1, 1, 1], [63, 1, 1, 1, 1], [127, 1, 1, 1, 1]],
 [[0, 1, 1, 0, 1], [4, 1, 1, 0, 1], [6, 1, 1, 0, 1], [14, 1, 1, 0, 1], [46, 1, 1, 0, 1], [47, 1, 1, 0, 1], [63, 1, 1, 0, 1], [127, 1, 1, 0, 1]]
]
```

Training_data is composed of ~60 such sequences some with 8 and some with fewer observations per sequence.

and then I train the model like this:
 `self.model = HiddenMarkovModel.from_samples(MultivariateGaussianDistribution, 8, training_data)`

It's not clear to me what's happening. But issues started to appear when I decided to normalize of `height` and `sleep_quality`. 

What seems strange is that sometimes this works, and other times the NaN error triggers or it trains but cant use `model.log_probability` for anything

**Comment by jmschrei** [comment]
Howdy

I'm guessing a reason that you get the error is because some of your columns always have the same value. This means that the standard deviation for a normal distribution will be 0 and you'll get weird results. I'd try excluding columns whose values don't change.

**Comment by precupstefan** [comment]
Hello,

Sorry for the late reply.

You are totally right. I fixed that by using a random value between 0-0.05 for 0 and 0.95-1 for 1.

Thanks for the help
